xbmc/profiles/ProfileManager.cpp

```diff
@@ -621,10 +621,12 @@ void CProfileManager::LoadMasterProfileForLogin()
     // determines that the (master) profile has only been loaded for login
     m_profileLoadedForLogin = true;
 
-    LoadProfile(0);
-
     // remember that the (master) profile has only been loaded for login
     m_previousProfileLoadedForLogin = true;
+
+    lock.unlock();
```

Comment:
I'm afraid the lock is held for a reason here and that releasing it will cause problems.

Reply:
In the tests it seems to work correctly and when switching profiles, the profiles and their contents are correct.

I've tried other ways to fix it but so far only this one works, any ideas are welcome

Reply:
I’m almost all cases is releasing a lock not the right fix. We need to understand the root cause. Can you provide stack traces where we can see the (at least) two involved mutexes ?

Sorry, not to understanding what’s going on and to „try“ if something happens is really very dangerous as a fix for a multithreading issue, because resulting problems caused by races are really very hard to debug and fix. 

I really world like not to merge this until we have a real understanding of the problem.

Reply:
Here is part of the backtrace of two threads trying to lock the same mutex (0x555559427ef8):

Thread 1:
```
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=<optimized out>, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ffff5c3c406 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff5c2287c in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ffff5c235a4 in __libc_message (fmt=fmt@entry=0x7ffff5db9f30 "Fatal glibc error: %s:%s (%s): assertion failed: %s\n") at ../sysdeps/posix/libc_fatal.c:150
#6  0x00007ffff5c339d7 in __libc_assert_fail (assertion=assertion@entry=0x7ffff5dbae80 "e != EDEADLK || (kind != PTHREAD_MUTEX_ERRORCHECK_NP && kind != PTHREAD_MUTEX_RECURSIVE_NP)", 
    file=file@entry=0x7ffff5db61d9 "pthread_mutex_lock.c", line=line@entry=445, function=function@entry=0x7ffff5dbea60 <__PRETTY_FUNCTION__.1> "__pthread_mutex_lock_full")
    at ./assert/__libc_assert_fail.c:31
#7  0x00007ffff5c921ad in __pthread_mutex_lock_full (mutex=0x555559427ef8) at ./nptl/pthread_mutex_lock.c:445
#8  0x00005555569925af in ADDON::CAddonMgr::GetAddon(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<ADDON::IAddon>&, ADDON::AddonType, ADDON::OnlyEnabled) const [clone .constprop.0] ()
#9  0x00005555566bb0da in KODI::GUILIB::GUIINFO::CAddonsGUIInfo::GetBool(bool&, CGUIListItem const*, int, KODI::GUILIB::GUIINFO::CGUIInfo const&) const ()
#10 0x00005555566b101e in KODI::GUILIB::GUIINFO::CGUIInfoProviders::GetBool(bool&, CGUIListItem const*, int, KODI::GUILIB::GUIINFO::CGUIInfo const&) const ()
...
#38 0x0000555556dbf380 in CProfileManager::FinalizeLoadProfile() ()
#39 0x0000555556dc3b4a in CProfileManager::LoadProfile(unsigned int) ()
#40 0x0000555556dc3ea3 in CProfileManager::LoadMasterProfileForLogin() ()
#41 0x0000555556dc3f78 in CProfileManager::LogOff() ()
...
#80 0x0000555556866302 in CApplication::Run() ()
#81 0x0000555555e5fb80 in main ()
```

Thread 45:
```
#1  0x00007ffff5c92226 in __pthread_mutex_lock_full (mutex=0x555559427ef8) at ./nptl/pthread_mutex_lock.c:441
#2  0x000055555698e1d9 in std::unique_lock<CCriticalSection>::lock() [clone .constprop.0] ()
#3  0x000055555699458d in ADDON::CAddonMgr::GetAddonsInternal(ADDON::AddonType, std::vector<std::shared_ptr<ADDON::IAddon>, std::allocator<std::shared_ptr<ADDON::IAddon> > >&, ADDON::OnlyEnabled, ADDON::CheckIncompatible) const ()
#4  0x000055555711b79e in XFILE::CAddonsDirectory::GetScriptsAndPlugins(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::shared_ptr<ADDON::IAddon>, std::allocator<std::shared_ptr<ADDON::IAddon> > >&) ()
#5  0x000055555711c6f6 in XFILE::CAddonsDirectory::GetScriptsAndPlugins(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, CFileItemList&) ()
#6  0x000055555712411f in XFILE::CAddonsDirectory::GetDirectory(CURL const&, CFileItemList&) ()
#7  0x0000555557143419 in XFILE::CDirectory::GetDirectory(CURL const&, std::shared_ptr<XFILE::IDirectory> const&, CFileItemList&, XFILE::CDirectory::CHints const&) ()
#8  0x0000555557144a07 in XFILE::CDirectory::GetDirectory(CURL const&, CFileItemList&, XFILE::CDirectory::CHints const&) ()
#9  0x0000555557144dd5 in XFILE::CDirectory::GetDirectory(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, CFileItemList&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int) ()
#10 0x0000555556661d7d in CDirectoryJob::DoWork() ()
#11 0x0000555556510e89 in CJobWorker::Process() ()
#12 0x00005555565cf0b5 in CThread::Action() ()
```

I've done a small test avoiding the call in one of the threads that tries to lock the mutex, but there are more locks, more analysis is needed.

Reply:
> Here is part of the backtrace of two threads trying to lock the same mutex

That's not what we are after. We need (at least) two mutexes. One thread owning mutex A while waiting to acquire mutex B. And another thread owning B while waiting for A.